In [2]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [6]:
dataset_path="C:/Users/Bunny/OneDrive/Documents/techworks/CNN/PetImages"

In [7]:
# Image preprocessing 
datagen = ImageDataGenerator(

    #normalize pixel 
    rescale=1./255, 
    
    # Data augmentation
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,

    #split data
    validation_split=0.2) 

In [8]:
# load training data
# Train data 

train_data = datagen.flow_from_directory(
    dataset_path,
    target_size=(128, 128),
    batch_size=32,
    class_mode='binary',
    subset='training'
)

Found 20000 images belonging to 2 classes.


In [9]:
# load validation data

val_data= datagen.flow_from_directory(
    dataset_path,
    target_size=(128, 128),
    batch_size=32,
    class_mode='binary',
    subset='validation'
)

Found 4998 images belonging to 2 classes.


In [10]:
print(train_data.class_indices)

{'Cat': 0, 'Dog': 1}


In [11]:
# Building the CNN model
model=keras.Sequential()

In [12]:
# First convolutional layer

model.add(
    keras.layers.Conv2D(
        filters=32, 
        kernel_size=(3, 3), 
        activation='relu', 
        input_shape=(128, 128, 3)
    )
)

c:\Users\Bunny\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [13]:
# First pooling layer

model.add(
    keras.layers.MaxPooling2D(pool_size=(2, 2))
)

In [14]:
# 2nd convolutional layer
model.add(
    keras.layers.Conv2D(
        filters=64, 
        kernel_size=(3, 3), 
        activation='relu'
    )
)

In [15]:
# 2nd pooling layer
model.add(
    keras.layers.MaxPooling2D(pool_size=(2, 2))
)

In [16]:
# Flattening layer

model.add(keras.layers.Flatten())

In [17]:
# Dense layer

model.add(
    keras.layers.Dense(
        units=128, activation='relu'
        ))

In [18]:
# output layer

model.add(
    keras.layers.Dense(
        units=1, activation='sigmoid'
    )
)

In [19]:
# view architecture
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 126, 126, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 63, 63, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 61, 61, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 30, 30, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 57600)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │     7,372,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 7,392,449 (28.20 MB)

 Trainable params: 7,392,449 (28.20 MB)

 Non-trainable params: 0 (0.00 B)

(None,126,126,Channels)

(Batch size,Height,width,No of Filters)

Input=128 * 128 * 3

filter=3*3

no of filters=32

output_size=((input-kernel)/stride)+1

stride=1

output_size=((128-3)/1)+1=125+1=126

height=126   width=126

filters=32

channels ---> 32

final_output=(None,126,126,32)

kernel

Formula=(kh * kh* InputChannels+1) * Filters

kernel=3 * 3

InputChannel: 3

Filters : 32

params=(3*3 * 3+1)*32

    =(27+1)*32

    =28*32
    
    =896


(None,126,126,Channels)

(Batch size,Height,width,No of Filters)

Input=128*128*3

filter=3*3

no of filters=32

output_size=((input-kernel)/stride)+1

stride=1

output_size=((128-3)/1)+1=125+1=126

height=126   width=126

filters=32

final_output=(None,126,126,32)

Kernel

Formula=(kernel_height*Kernel_width*Input_Channles+1)*Filters

kernels=3*3

Input_channels=3

Filters=32

params=(3 * 3 * 3 + 1) * 32
      
      =(27+1) * 32 = 896

### Pooling layer

MaxPooling((2,2))

pooling=No learning,No weights

(only reuces size)

Input=126*126*32

pool=2*2

Formula=126/2

  =63

Output:63*63*32

Final:(None,63,63,32)

### Second Convolution layer

Input from pooling : (None,63,63,32)

kernel = 3*3 

output = (63-3+1) = 61

(None,61,61,32)

params=9248

=(3 * 3 * 32 + 1) * 32

=289*32

=9248

### Second Pooling

(None,30,30,32)

Param=0

input=(61,61,32)

Pooling=2*2

Floor divison = 61/2 = 30

Final:(None,30,30,32)

No Parameters.







In [21]:
import tensorflow as tf
print(tf.config.list_physical_devices())


# Compile CNN
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)




[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]


In [23]:
# Train the model
history=model.fit(
    train_data,
    epochs=2,
    validation_data=val_data
)

Epoch 1/2
344/625 ━━━━━━━━━━━━━━━━━━━━ 2:11 470ms/step - accuracy: 0.6708 - loss: 0.6091

c:\Users\Bunny\anaconda3\Lib\site-packages\PIL\TiffImagePlugin.py:949: UserWarning: Truncated File Read
  warnings.warn(str(msg))


625/625 ━━━━━━━━━━━━━━━━━━━━ 314s 503ms/step - accuracy: 0.6914 - loss: 0.5884 - val_accuracy: 0.7393 - val_loss: 0.5285
Epoch 2/2
625/625 ━━━━━━━━━━━━━━━━━━━━ 241s 385ms/step - accuracy: 0.7530 - loss: 0.5120 - val_accuracy: 0.7471 - val_loss: 0.5235


In [24]:
loss, accuracy = model.evaluate(val_data)
print(f"Validation Loss: {loss:.4f}")
print(f"Validation Accuracy: {accuracy:.4f}")

157/157 ━━━━━━━━━━━━━━━━━━━━ 31s 199ms/step - accuracy: 0.7497 - loss: 0.5153
Validation Loss: 0.5153
Validation Accuracy: 0.7497
